internal/querycoord/replica.go

```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
 		}
 	}
 }
+
+// ApplyBalancePlan applies balancePlan to replica nodes.
+func (rep *ReplicaInfos) ApplyBalancePlan(p *balancePlan, kv kv.MetaKv) error {
+	rep.globalGuard.Lock()
+	defer rep.globalGuard.Unlock()
+
+	var sourceReplica, targetReplica *milvuspb.ReplicaInfo
+
+	// check source and target replica ids are valid
+	if p.sourceReplica != invalidReplicaID {
+		replica, ok := rep.replicas[p.sourceReplica]
+		if !ok {
+			return errors.New("replica not found")
+		}
+		sourceReplica = proto.Clone(replica).(*milvuspb.ReplicaInfo)
+	}
+
+	if p.targetReplica != invalidReplicaID {
+		replica, ok := rep.replicas[p.targetReplica]
+		if !ok {
+			return errors.New("replica not found")
+		}
+		targetReplica = proto.Clone(replica).(*milvuspb.ReplicaInfo)
+	}
+
+	var replicasChanged []*milvuspb.ReplicaInfo
+
+	// generate ReplicaInfo to save to MetaKv
+	if sourceReplica != nil {
+		// remove node from replica node list
+		removeNodeFromReplica(sourceReplica, p.nodeID)
+		replicasChanged = append(replicasChanged, sourceReplica)
+	}
+	if targetReplica != nil {
+		// add node to replica
+		targetReplica.NodeIds = append(targetReplica.NodeIds, p.nodeID)
+		replicasChanged = append(replicasChanged, targetReplica)
+	}
+
+	// save to etcd first
+	if len(replicasChanged) > 0 {
```

Comment:
it can be simple to implement these in MetaReplica, which allows you to use the setReplica() method.

Reply:
This [example](https://github.com/milvus-io/milvus/pull/17218/files#diff-ade465f7861613623d78a92764a878a2fcf296fccc1f260a64f3829a2b3b006dR1303) may help

Reply:
@yah01 tried this way but cannot keep this whole operation atomic.